src/cosmos/engine/CosmosEngine.ts

```diff
@@ -286,53 +285,66 @@ export class CosmosEngine extends CurrencyEngine<
-            const swapTxData = asCosmosSwapTransactionDataFromJson(txData)
-
-            let gasFeeCoin = coin('0', this.networkInfo.nativeDenom)
-            let gasLimit = '200000' // Default gas limit for swaps
-            let networkFee = '0'
-            let swapMemo: string | undefined
+            // Accept only Rango provider-shaped payload for DEX swaps
+            const rango = asRangoProviderTxDataFromJson(txData)
 
-            // Extract memo - critical for cross-chain routing
-            if (swapTxData.memo != null && swapTxData.memo !== '') {
-              swapMemo = swapTxData.memo
-              memos.push({ type: 'text', value: swapMemo })
-              this.log(`DEX swap memo: ${swapMemo}`)
+            // Build TxBody from Rango protoMsgs or msgs
+            let messages: EncodeObject[] = []
+            if (rango.protoMsgs != null && rango.protoMsgs.length > 0) {
+              messages = rango.protoMsgs.map((m: any) => {
+                const valueBytes = coerceToBytes(m.value)
+                if (valueBytes == null) throw new Error('Rango txData: message value is not bytes')
+                return { typeUrl: String(m.type_url), value: valueBytes }
+              })
+            } else if (rango.msgs != null && rango.msgs.length > 0) {
+              const aminoTypes = new AminoTypes({
+                ...createBankAminoConverters(),
+                ...createIbcAminoConverters()
+              })
+              messages = rango.msgs.map((aminoMsg: any) => aminoTypes.fromAmino(aminoMsg))
+            } else {
+              throw new Error('Rango txData has no messages')
             }
 
-            // Extract fee information from swap transaction data
-            if (swapTxData.fee.amount.length > 0) {
-              const feeAmount = swapTxData.fee.amount[0]
-              networkFee = feeAmount.amount
-              gasFeeCoin = coin(networkFee, feeAmount.denom)
+            // Estimate gas and compute fee using chain gas price (high) for safety
+            const { queryClient } = this.getClients()
+            const { gasInfo } = await queryClient.tx.simulate(
+              messages,
+              rango.memo,
+              encodeSecp256k1Pubkey(base16.parse(this.walletInfo.keys.publicKey)),
+              this.getSequence()
+            )
+            if (gasInfo?.gasUsed == null) throw new Error(`simulate didn't return gasUsed `)
+            const gasLimit = ceil(mul(gasInfo.gasUsed.toString(), '1.5'), 0)
+            const { high } = getGasPriceRangesFromChain(this.tools.chainData)
+            const feeDenom = this.networkInfo.nativeDenom
+            const gasFee = ceil(mul(gasLimit, high.toString()), 0)
+            const gasFeeCoin = coin(gasFee, feeDenom)
+            const networkFee = gasFeeCoin.amount
+
+            // Memo (for routing/debugging)
+            if (rango.memo != null && rango.memo !== '') {
+              memos.push({ type: 'text', value: rango.memo })
```

Comment:
Since the cleaner is `memo: asOptional(asString, '')`, the undefined case should never happen. You can either get rid of the `!= null` check, or get rid of the `asOptional(, '')` fallback value in the cleaner above.